### PR TITLE
♻️ Relax condition on modifiers

### LINKED ISSUE #1678: 🐛 Operation folds are not modifier-aware, leaving modifier body in unexpected state after fold to identity
### System and Environment Information

- clang/mlir/llvm version 22.1.0-rc3
- OS: NixOS 26.05 (Yarara), kernel 6.19.10-cachyos, x86_64
- MQT Core: main (54c30168a997b1fd6ddf118cd3ea2b7607532723)

### Bug Description

Folds on QCO operations (e.g. `POp::fold`) are not aware that the op may live inside a modifier body (e.g. `qco.ctrl`). When a fold reduces the op to identity, the modifier body is left containing only `qco.yield` - an unexpected state, causing a crash.

### Steps to Reproduce

Try to compile this with `mqt-cc` (will segfault):
``` mlir
module {
    func.func @controlled-zero-phase() {
      %ctrl = qco.alloc : !qco.qubit
      %tgt  = qco.alloc : !qco.qubit
      %cst  = arith.constant 0.0 : f64
      %ctrl_1, %tgt_1 = qco.ctrl(%ctrl) targets(%t = %tgt) {
        %t_1 = qco.p(%cst) %t : !qco.qubit -> !qco.qubit
        qco.yield %t_1
      } : ({!qco.qubit}, {!qco.qubit}) -> ({!qco.qubit}, {!qco.qubit})
      qco.sink %ctrl_1 : !qco.qubit
      qco.sink %tgt_1  : !qco.qubit
      return
    }
  }
```

### LINKED ISSUE #1727: ✨ Relax condition on modifiers
#### Description

Currently, the modifiers in QC and QCO enforce that their bodies contain a single unitary operation followed by a `yield` operation. This has led to more complications than benefits. The condition should be dropped altogether, and arbitrarily many operations (unitary as well as, e.g., `arith.constant`s) should be legal.

#### Acceptance criteria

- The condition is dropped.
- All relevant canonicalization patterns are updated accordingly.
- The translations from `qc::QuantumComputation` and OpenQASM to QC are updated accordingly. In particular, `qc::CompoundOperation`s inside modifiers can be translated, and the test case mentioned in https://github.com/munich-quantum-toolkit/core/pull/1717#discussion_r3262539677 is added.
- Tests are added to cover new edge cases.

### PR DESCRIPTION
**Pull request by denialhaag**
## Description

This PR relaxes the condition that modifiers in QC and QCO must have a single unitary operation followed by a `yield` operation. Modifiers may now contain an arbitrary number of operations (unitary as well as, e.g., `arith.constant`s).

Fixes #1678
Fixes #1727

## Checklist

- [x] The pull request only contains commits that are focused and relevant to this change.
- [x] I have added appropriate tests that cover the new/changed functionality.
- [x] I have updated the documentation to reflect these changes.
- [x] I have added entries to the changelog for any noteworthy additions, changes, fixes, or removals.
- [x] ~~I have added migration instructions to the upgrade guide (if needed).~~
- [x] The changes follow the project's style guidelines and introduce no new warnings.
- [x] The changes are fully tested and pass the CI checks.
- [x] I have reviewed my own code changes.

### CODE REVIEW

**Comment by denialhaag** [comment]
@coderabbitai full review

File: mlir/lib/Conversion/QCOToJeff/QCOToJeff.cpp [hunk 1]
```diff
@@ -934,6 +943,13 @@ struct ConvertQCOInvOpToJeff final : StatefulOpConversionPattern<InvOp> {
     state.invOp = op;
     if (state.targetsIn.empty()) {
       state.targetsIn = llvm::to_vector(adaptor.getQubitsIn());
+    } else {
+      auto outerQubits = state.targetsIn;
+      SmallVector<Value> innerQubits;
+      for (auto arg : op.getBody()->getArguments()) {
+        innerQubits.push_back(outerQubits[arg.getArgNumber()]);
+      }
+      state.targetsIn = std::move(innerQubits);
```
[the remark this answers is not in the document]
Comment (denialhaag) [reply mlir/lib/Conversion/QCOToJeff/QCOToJeff.cpp:948]: The conversion explicitly fails if `qco.inf` is nested inside `qco.ctrl`. This is averted by running the canonicalization pass first.

File: mlir/lib/Dialect/QC/Translation/TranslateQuantumComputationToQC.cpp [hunk 2]
```diff
@@ -546,6 +576,60 @@ static LogicalResult translateOperation(QCProgramBuilder& builder,
...
+      }
+      for (const auto& op : compoundOp) {
+        if (failed(translateOperation(builder, *op, state))) {
+          llvm::reportFatalInternalError("Failed to translate operation inside "
+                                         "controlled CompoundOperation");
+        }
+      }
+      state.ctrlTargets.clear();
+      state.inCtrlOp = false;
+    });
```
[the remark this answers is not in the document]
Comment (denialhaag) [reply mlir/lib/Dialect/QC/Translation/TranslateQuantumComputationToQC.cpp:644]: `targets` will always be built from `sortedPairs`. It's the only reason we define `sortedPairs` in the first place.

**Comment by denialhaag** [comment]
@coderabbitai review

File: mlir/lib/Conversion/QCOToJeff/QCOToJeff.cpp [hunk 3]
```diff
@@ -871,18 +880,24 @@ struct ConvertQCOCtrlOpToJeff final : StatefulOpConversionPattern<CtrlOp> {
   LogicalResult
   matchAndRewrite(CtrlOp op, OpAdaptor adaptor,
                   ConversionPatternRewriter& rewriter) const override {
+    if (op.getNumBodyUnitaries() != 1) {
+      return rewriter.notifyMatchFailure(
+          op,
+          "Control modifiers with multiple body unitaries are not supported.");
+    }
```
[the remark this answers is not in the document]
Comment (denialhaag) [reply mlir/lib/Conversion/QCOToJeff/QCOToJeff.cpp:871]: Yeah, this will be added in a follow-up. We plan to add a pass that unrolls `qco::CtrlOp`s. Once unrolled, it will be much easier to convert to `jeff` because `jeff` does not have the concepts of modifiers. `jeff` represents controlled gates via operation attributes.

[the remark this answers is not in the document]
Comment (denialhaag) [reply mlir/lib/Conversion/QCOToJeff/QCOToJeff.cpp:871]: No, I'll open the issue myself later today. 🙂

[the remark this answers is not in the document]
Comment (denialhaag) [reply mlir/lib/Conversion/QCOToJeff/QCOToJeff.cpp:871]: See https://github.com/munich-quantum-toolkit/core/issues/1758.

**Comment by denialhaag** [comment]
@coderabbitai review

**Comment by denialhaag** [comment]
@coderabbitai review

**Comment by burgholzer** [comment]
I am still working on this. I just pushed a first batch of changes to get a better PR diff again.

**Review by burgholzer — APPROVED** [review]
Alright. I am through with all of the changes here.
This was a big one. It took a couple of hours to go through everything.
I managed to simplify a couple of concepts here and there while working my way through. Conceptually, this was pretty clean already.

Please have another close look at the changes that I introduced and see whether they make sense. Some of them are pretty nice simplifications I believe.
This is good to go from my side.

**Comment by burgholzer** [comment]
@coderabbitai review

File: mlir/lib/Dialect/QC/IR/Modifiers/CtrlOp.cpp [hunk 4]
```diff
@@ -33,23 +37,42 @@ struct MergeNestedCtrl final : OpRewritePattern<CtrlOp> {
...
+    }
+    const auto targets =
+        llvm::map_to_vector(innerCtrlOp.getTargets(), [&](Value t) {
+          return utils::getValueFromBlockArgument(t, outerTargets);
+        });
+
+    auto merged = CtrlOp::create(rewriter, op.getLoc(), controls, targets);
+    rewriter.inlineRegionBefore(innerCtrlOp.getRegion(), merged.getRegion(),
+                                merged.getRegion().end());
+    rewriter.eraseOp(op);
```
[the remark this answers is not in the document]
Comment (burgholzer) [reply mlir/lib/Dialect/QC/IR/Modifiers/CtrlOp.cpp:81]: While this might be true in theory, I do not see the practical case where this would be coming up. If the current operation is a CtrlOp and it contains a single unitary that is another CtrlOp. There is no reason to believe that there would be other operations in the region of the outside CtrlOp.

We may want to add a sanity check that points this out though.

What would you propose as an alternative solution if the goal is to carry/remap the full outer body through the rewrite.

**Comment by denialhaag** [comment]
@coderabbitai review

**Comment by denialhaag** [comment]
Thanks a lot for all the improvements, @burgholzer! I really like how clean this PR has become despite its size. Given that CodeRabbit is happy as well, I'll set this to auto-merge. 🎉
